src/customfilesystemmodel.cpp

```diff
@@ -82,27 +104,49 @@ bool CustomFileSystemModel::setData(const QModelIndex &index,
             _checklist.insert(index);
             emit dataChanged(index, index, selectionChangedRole);
 
-            QModelIndex parent = index.parent();
-            if(parent.isValid())
+            // Recursively set PartiallyChecked on all ancestors, and make an
+            // ordered list of checked ancestors.
+            QModelIndex        parent         = index.parent();
+            QModelIndex        previousParent = index;
+            QList<QModelIndex> checkedAncestors;
+            while(parent.isValid())
+            {
+                if((dataInternal(parent) == Qt::Checked) ||
+                        hasCheckedAncestor(parent))
+                    checkedAncestors << parent;
+
+                // Set parent to be PartiallyChecked.
+                setIndexCheckState(parent, Qt::PartiallyChecked);
+
+                // Ascend to higher level of ancestor.
+                previousParent = parent;
```

Comment:
The changes are good, however do you really need previousParent in this first block? I think you only need it further down for the sibling iteration.